src/preprocessing/passes/normalize.cpp

```diff
@@ -0,0 +1,1063 @@
+              {
+                if (patA[j] != patB[j])
+                {
+                  return patA[j] < patB[j];
+                }
+              }
+
+              // If one segment has more non-negative values, it is considered
+              // greater
+              if (patA.size() != patB.size())
+              {
+                return patA.size() < patB.size();
+              }
+            }
+
+            ++itA;
+            ++itB;
+          }
+
+          // Handle cases where one iterator reaches the end before the other
+          if (itA != a->varNames.end()) return true;
+          if (itB != b->varNames.end()) return false;
+
+          return false;
+        });
+  }
+
+  // ----------------------------------------
+  // Step 6: Collect and normalize types
+  // ----------------------------------------
+  std::vector<TypeNode> types;
+  std::unordered_set<TNode> visited;
+  std::unordered_set<TypeNode> mark;
+  for (const std::vector<NodeInfo*>& eqClass : eqClasses)
+  {
+    for (NodeInfo* nodeInfo : eqClass)
+    {
+      collectTypes(nodeInfo->node, types, visited, mark);
+    }
+  }
+
+  std::unordered_map<TypeNode, TypeNode> normalizedSorts;
+  size_t sortCounter = 0;
+  for (const TypeNode& ctn : types)
+  {
+    if (ctn.isUninterpretedSort() && ctn.getNumChildren() == 0)
+    {
+      std::string new_sort_name = "S"
+                                  + std::string(8 - numDigits(sortCounter), '0')
+                                  + std::to_string(sortCounter);
+      sortCounter++;
+      TypeNode new_sort = nodeManager()->mkSort(new_sort_name);
+      normalizedSorts[ctn] = new_sort;
+    }
+  }
+
+  NormalizeSortNodeConverter* sortNormalizer =
+      new NormalizeSortNodeConverter(normalizedSorts, nodeManager());
+
+  // ----------------------------------------
+  // Step 7: Normalize nodes based on sorted order
+  // ----------------------------------------
+  std::unordered_map<Node, Node> freeVar2node;
+  std::unordered_map<Node, Node> boundVar2node;
+  std::unordered_map<std::string, std::string> normalizedName;
+
+  bool hasQID = false;
+  for (const auto& eqClass : eqClasses)
+  {
+    for (const auto& ni : eqClass)
+    {
+      Node renamed = rename(ni->node,
+                            freeVar2node,
+                            boundVar2node,
+                            normalizedName,
+                            normalizedSorts,
+                            nodeManager(),
+                            d_preprocContext,
+                            sortNormalizer,
+                            hasQID);
+      ni->node = renamed;
+    }
+  }
+
+  // ----------------------------------------
+  // Step 8: Sort nodes within equivalence classes based on normalized names
+  // ----------------------------------------
+  for (auto& eqClass : eqClasses)
+  {
+    std::sort(eqClass.begin(),
+              eqClass.end(),
+              [&normalizedName](NodeInfo* a, NodeInfo* b) {
```

Comment:
Ditto, it'd be better if this was a utility method instead of defined in place.

Reply:
Changed